### ISSUE TITLE
TsTv by Qual for VCFTools breaks report on infinity values

### ISSUE THREAD

**Issue by alneberg** [ewels/MultiQC#894]
**Description of bug:**
Running MultiQC on the attached TsTv.qual file from VCFtools generates faulty json, which causes a javascript exception. It is likely due to the `inf` or `-nan` values found within the file. The file is generated using artificially small test-data, so this error might not occur in reality.

The exception is caused by the term `Infinity` within the json data block.

**MultiQC Error log:**
Multiqc log (no error): 
[multiqc.log](https://github.com/ewels/MultiQC/files/2808108/multiqc.log)

Error in browser:
``` javascript
multiqc_report.html:899 jQuery.Deferred exception: Unexpected token I in JSON at position 351 SyntaxError: Unexpected token I in JSON at position 351
    at JSON.parse (<anonymous>)
    at HTMLDocument.<anonymous> (file:///Users/johannes.alneberg/Downloads/reproducing_multiqc/multiqc_report.html:1823:20)
    at j (file:///Users/johannes.alneberg/Downloads/reproducing_multiqc/multiqc_report.html:899:29948)
    at k (file:///Users/johannes.alneberg/Downloads/reproducing_multiqc/multiqc_report.html:899:30262) undefined
r.Deferred.exceptionHook @ multiqc_report.html:899
multiqc_report.html:899 Uncaught SyntaxError: Unexpected token I in JSON at position 351
    at JSON.parse (<anonymous>)
    at HTMLDocument.<anonymous> (multiqc_report.html:1823)
    at j (multiqc_report.html:899)
    at k (multiqc_report.html:899)
```

The unformatted json:
``` json
"{"vcftools_tstv_by_qual": {"plot_type": "xy_line", "datasets": [[{"data": [[16.86, 2.16667], [21.8, 2.6], [41.77, 2.2], [50.77, 2.0], [69.77, 1.8], [106.77, 1.6], [111.77, 2.0], [124.03, 1.75], [137.77, 1.5], [140.77, 2.0], [150.77, 1.66667], [226.77, 1.33333], [236.77, 1.0], [260.77, 0.666667], [318.77, 1.0], [326.77, 0.5], [352.77, 1.0], [370.77, Infinity], [419.77, null]], "name": "haplotypecaller_9877R.TsTv.qual"}]], "config": {"xlab": "SNP Quality Threshold", "smooth_points_sumcounts": false, "smooth_points": 400, "title": "VCFTools: TsTv by Qual", "xmin": 0, "ylab": "TsTv Ratio", "ymin": 0, "id": "vcftools_tstv_by_qual"}}}"
```
Formatted json:
``` json
LZString.decompressFromBase64(mqc_compressed_plotdata)
"{"vcftools_tstv_by_qual": {
    "plot_type": "xy_line", 
    "datasets": [[{
        "data": [ 
             [16.86, 2.16667], [21.8, 2.6], [41.77, 2.2], [50.77, 2.0], [69.77, 1.8], 
             [106.77, 1.6], [111.77, 2.0], [124.03, 1.75], [137.77, 1.5], [140.77, 2.0], 
             [150.77, 1.66667], [226.77, 1.33333], [236.77, 1.0], [260.77, 0.666667], 
             [318.77, 1.0], [326.77, 0.5], [352.77, 1.0], [370.77, Infinity], 
             [419.77, null]
        ], 
        "name": "haplotypecaller_9877R.TsTv.qual"
        }]], 
    "config": {
        "xlab": "SNP Quality Threshold", 
        "smooth_points_sumcounts": false, 
        "smooth_points": 400, 
        "title": "VCFTools: TsTv by Qual", 
        "xmin": 0, 
        "ylab": "TsTv Ratio", 
        "ymin": 0, 
        "id": "vcftools_tstv_by_qual"
      }
}}"
```

**File that triggers the error:**
[haplotypecaller_9877R.TsTv.qual.txt](https://github.com/ewels/MultiQC/files/2807933/haplotypecaller_9877R.TsTv.qual.txt)

**MultiQC run details (please complete the following):**
 - Command used to run MultiQC: `multiqc -f -v .`
 - MultiQC Version: MultiQC v1.7
 - Operating System: macOS High Sierra 10.13.6
 - Python Version: Python 3.6.5
 - Method of MultiQC installation: pipenv

**Comment by ewels** [comment]
Two fix options:

1) Force variable type in VCFtools parsing code, here:
https://github.com/ewels/MultiQC/blob/2ef19050e1f734c7e19651f09ae88082d3a7a5ef/multiqc/modules/vcftools/tstv_by_qual.py#L22

(eg. `d[key] = float(val)` and handle exceptions)

2) Implement more general cleanup in the linegraph code, again enforcing variables types here:
https://github.com/ewels/MultiQC/blob/2ef19050e1f734c7e19651f09ae88082d3a7a5ef/multiqc/plots/linegraph.py#L118

I'm kind of surprised that this isn't already done to be honest. Should always be able to convert the `y` variable to a `float` and the `x` can also be converted if `categories` isn't set.

Phil

**Comment by alneberg** [comment]
Thanks for the suggestions! 

Actually in your first suggestion, the `float()`-conversion is already done for both `key` and `val` in the two rows above:
https://github.com/ewels/MultiQC/blob/2ef19050e1f734c7e19651f09ae88082d3a7a5ef/multiqc/modules/vcftools/tstv_by_qual.py#L20-L22

So it turns out, `'inf'` is actually a valid input to the function `float`. But I will debug it further to see it goes bad somewhere else in the stack.

Even though it should be technically easy to fix this, I'm not sure what the best substitution for these strange values would be? Or would the plotting tool be able to handle infinite values as long as it's properly formatted?

**Comment by ewels** [comment]
I think you can just use `null` Javascript values. Which I guess is `None` in Python before being dumped to JSON.

eg. https://www.highcharts.com/demo/area-missing

**Comment by ewels** [comment]
Thanks @alneberg!

**Comment by ewels** [comment]
Added a more general fix in 0a2e583 which just removes `Infinity` from the dumped JSON string.
